### PR TITLE
[mcs] Add platforms to profile

### PR DESCRIPTION
**Pull request by luhenry**

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -4554,7 +4554,7 @@ fi
     echo "MONO_CORLIB_VERSION = $MONO_CORLIB_VERSION" >> $mcs_topdir/build/config.make
```
Comment (vargaz) [inline configure.ac]: Why is this necessary ? People have been using PROFILE for years, ie. make all PROFILE=net_4_x. There might be build scripts etc lying around doing the same.

File: mcs/build/rules.make [hunk 2]
```diff
@@ -25,18 +25,16 @@ Q=$(if $(V),,@)
...
 USE_MBAS_FLAGS = /codepage:$(CODEPAGE) $(LOCAL_MBAS_FLAGS) $(PLATFORM_MBAS_FLAGS) $(PROFILE_MBAS_FLAGS) $(MBAS_FLAGS)
 USE_CFLAGS = $(LOCAL_CFLAGS) $(CFLAGS) $(CPPFLAGS)
-CSCOMPILE = $(Q_MCS) $(MCS) $(USE_MCS_FLAGS)
-CSC_RUNTIME_FLAGS = --aot-path=$(abspath $(topdir)/class/lib/$(BUILD_TOOLS_PROFILE)) --gc-params=nursery-size=64m
+CSCOMPILE = $(Q_MCS) $(BUILDMCS) $(USE_MCS_FLAGS)
+CSC_RUNTIME_FLAGS = --aot-path=$(abspath $(topdir)/class/lib/$(BUILDPROFILE)) --gc-params=nursery-size=64m
 BASCOMPILE = $(MBAS) $(USE_MBAS_FLAGS)
 CCOMPILE = $(CC) $(USE_CFLAGS)
-BOOT_COMPILE = $(Q_MCS) $(BOOTSTRAP_MCS) $(USE_MCS_FLAGS)
+BOOT_COMPILE = $(Q_MCS) $(BUILDMCS) $(USE_MCS_FLAGS)
```
Comment (marek-safar) [inline mcs/build/rules.make]: This is wrong build != bootstrap

File: mcs/class/System.Data/Test/DataProviderTests/datareadertests/Makefile [hunk 3]
```diff
@@ -13,10 +13,10 @@ NO_SIGN_ASSEMBLY=yes
 include $(topdir)/build/library.make
 
 Runtests.exe: $(the_lib)
-	$(MCS) /r:$(the_lib) Runtests.cs
+	$(BUILDMCS) /r:$(the_lib) Runtests.cs
 
 DbRetrieve.dll: 
-	$(MCS) /r:$(REFERENCES) /target:library /out:DbRetrieve.dll @$(SOURCEFILE)
```
Comment (marek-safar) [inline mcs/class/System.Data/Test/DataProviderTests/datareadertests/Makefile]: If you want to change this then `CSCOMPILE` is correct value

File: mcs/build/profiles/build.make [hunk 4]
```diff
@@ -1,10 +1,6 @@
 # -*- makefile -*-
 
-BOOTSTRAP_PROFILE = basic
-BUILD_TOOLS_PROFILE = basic
-
-BOOTSTRAP_MCS = MONO_PATH="$(topdir)/class/lib/$(BOOTSTRAP_PROFILE)$(PLATFORM_PATH_SEPARATOR)$$MONO_PATH" $(INTERNAL_CSC)
-MCS = $(BOOTSTRAP_MCS)
+BUILDPROFILE = basic
```
Comment (marek-safar) [inline mcs/build/profiles/build.make]: I think BUILD_TOOLS_PROFILE is much better name

File: runtime/Makefile.am [hunk 5]
```diff
@@ -26,46 +26,57 @@ $(symlinks):
 SUPPORT_FILES = $(symlinks) mono-wrapper etc/mono/config
 
 build_profiles = 
+build_platforms = 
 
 if INSTALL_4_x
 build_profiles += binary_reference_assemblies net_4_x xbuild_12 xbuild_14
+build_platforms += macos win32 linux
```
Comment (marek-safar) [inline runtime/Makefile.am]: Why do you need this, seems like just duplicate work. We also need to do `make` in mono/mcs and it needs to work just like today

File: mcs/Makefile [hunk 6]
```diff
@@ -26,6 +26,28 @@ orbis_SUBDIRS := build class
 
 include build/rules.make
 
+monodroid_PLATFORMS := android
```
Comment (marek-safar) [inline mcs/Makefile]: This should go to profile config file e.g. `build/profiles/monotouch.make`

File: mcs/Makefile [hunk 7]
```diff
@@ -26,6 +26,28 @@ orbis_SUBDIRS := build class
...
+monotouch_PLATFORMS := ios
+monotouch_watch_PLATFORMS := watchos
+monotouch_tv_PLATFORMS := tvos
+monotouch_runtime_PLATFORMS := macos
+monotouch_watch_runtime_PLATFORMS := macos
+monotouch_tv_runtime_PLATFORMS := macos
+xammac_PLATFORMS := macos
+testing_aot_hybrid_PLATFORMS := macos win32 linux
+testing_aot_full_PLATFORMS := macos win32 linux
+binary_reference_assemblies_PLATFORMS := macos win32 linux
```
Comment (marek-safar) [inline mcs/Makefile]: Don't need it here

File: mcs/Makefile [hunk 8]
```diff
@@ -26,6 +26,28 @@ orbis_SUBDIRS := build class
...
+net_4_x_PLATFORMS := macos win32 linux
+xammac_net_4_5_PLATFORMS := macos
+xbuild_12_PLATFORMS := macos win32 linux
```
Comment (marek-safar) [inline mcs/Makefile]: Don't need it here

File: mcs/Makefile [hunk 9]
```diff
@@ -26,6 +26,28 @@ orbis_SUBDIRS := build class
...
+xbuild_14_PLATFORMS := macos win32 linux
```
Comment (marek-safar) [inline mcs/Makefile]: Don't need it here

File: mcs/build/rules.make (hunk 2 above)
Comment (luhenry) [inline mcs/build/rules.make]: BOOTSTRAP_MCS and BUILD_TOOLS_PROFILE always have the same value.

File: mcs/Makefile (hunk 7 above)
Comment (luhenry) [inline mcs/Makefile]: That's a leftover, I will remove it.

File: mcs/Makefile (hunk 8 above)
Comment (luhenry) [inline mcs/Makefile]: Same as https://github.com/mono/mono/pull/4764#discussion_r114134417

File: mcs/Makefile (hunk 9 above)
Comment (luhenry) [inline mcs/Makefile]: Same as https://github.com/mono/mono/pull/4764#discussion_r114134417

File: mcs/Makefile (hunk 6 above)
Comment (luhenry) [inline mcs/Makefile]: I will move it to `build/profiles/$(PROFILE).make`. For profiles that only run on 1 platform, it just makes it much easier to still provide the 1 platform, so we can generalise the rest of the build system, and not special case for when the platform is provided or not.
It also opens the door (at a  later time) for what @kumpera suggested of generalising profiles (for example: orbis becomes monotouch + orbis platform or something equivalent).

File: runtime/Makefile.am (hunk 5 above)
Comment (luhenry) [inline runtime/Makefile.am]: Moving the platforms to `build/profiles/$(PROFILE).make` will fix that.
